File: src/servicebroker-npm/test/multiplexingRelayServiceBrokerTests.ts
import { RemoteServiceConnections } from '../src/constants'
import { FrameworkServices } from '../src/FrameworkServices'
import { IDisposable } from '../src/IDisposable'
import { IRemoteServiceBroker } from '../src/IRemoteServiceBroker'
import { IServiceBroker } from '../src/IServiceBroker'
import { MultiplexingRelayServiceBroker } from '../src/MultiplexingRelayServiceBroker'
import { FullDuplexStream, MultiplexingStream } from 'nerdbank-streams'
import { RemoteServiceBroker } from '../src/RemoteServiceBroker'
import { ICalculatorService } from './testAssets/interfaces'
import { MockServiceBroker } from './testAssets/mockServiceBroker'
import { calcDescriptorMsgPackBE32, calcDescriptorUtf8Http } from './testAssets/testUtilities'
import CancellationToken from 'cancellationtoken'
import { BrokeredServicesChangedArgs } from '../src/BrokeredServicesChangedArgs'

describe('MultiplexingRelayServiceBroker', function () {
	let innerServer: IServiceBroker

	beforeEach(function () {
		innerServer = new MockServiceBroker()
	})

	describe('connectToServer', function () {
		it('success case', async function () {
			const client = await getRelayClient()
			await assertCalculatorService(client)
		})

		it('canceled', async function () {
			const pair = FullDuplexStream.CreatePair()
			await expect(() =>
				MultiplexingRelayServiceBroker.connectToServer(new MockServiceBroker(), pair.first, CancellationToken.CANCELLED)
			).rejects.toThrowError(CancellationToken.CancellationError)
		})
	})

	describe('constructor', function () {
		it('owned mxstream', async function () {
			const pair = FullDuplexStream.CreatePair()
			const clientTask = RemoteServiceBroker.connectToMultiplexingDuplex(pair.first)
			const serverTask = MultiplexingStream.CreateAsync(pair.second)
			const serverMxStream = await serverTask
			const relayChannel = await serverMxStream.offerChannelAsync('')
			const relay = new MultiplexingRelayServiceBroker(innerServer, serverMxStream, true)
			FrameworkServices.remoteServiceBroker.constructRpc(relay, relayChannel)

			await assertCalculatorService(await clientTask)
		})
	})

	describe('requestServiceChannel', function () {
		it('non-existent service', async function () {
			const client = await getRelayClient()
			const pipe = await client.getPipe(calcDescriptorMsgPackBE32.moniker)
			expect(pipe).toBeNull()
		})
	})

	describe('cancelServiceRequest', function () {
		it('cancels channel offer', async function () {
			const pair = FullDuplexStream.CreatePair()
			const relayTask = MultiplexingRelayServiceBroker.connectToServer(innerServer, pair.second)
			const clientMx = await MultiplexingStream.CreateAsync(pair.first)
			const clientChannel = await clientMx.acceptChannelAsync('')
			await relayTask

			const clientBrokerProxy = FrameworkServices.remoteServiceBroker.constructRpc<IRemoteServiceBroker>(clientChannel)
			const connectionInfo = await clientBrokerProxy.requestServiceChannel(calcDescriptorUtf8Http.moniker)
			await clientBrokerProxy.cancelServiceRequest(connectionInfo.requestId!)

			// Assert that the offer for the calculator channel is rescinded.
			expect(() => clientMx.acceptChannel(connectionInfo.multiplexingChannelId!)).toThrow()
		})
	})

	describe('completion', function () {
		it('client closed', async function () {
			const pair = FullDuplexStream.CreatePair()
			const clientTask = RemoteServiceBroker.connectToMultiplexingDuplex(pair.first)
			const serverTask = MultiplexingStream.CreateAsync(pair.second)
			const serverMxStream = await serverTask
			const relayChannel = await serverMxStream.offerChannelAsync('')
			const relay = new MultiplexingRelayServiceBroker(innerServer, serverMxStream, true)
			FrameworkServices.remoteServiceBroker.constructRpc(relay, relayChannel)
			relayChannel.dispose()
			await expect(clientTask).rejects.toThrow()
			await relay.completion
		})
	})

	describe('handshake', function () {
		it('without multiplexing', async function () {
			const client = await getRemoteClientProxy()
			await expect(client.handshake({ supportedConnections: RemoteServiceConnections.IpcPipe })).rejects.toThrow()
		})

		it('with multiplexing', async function () {
			const client = await getRemoteClientProxy()
			await client.handshake({ supportedConnections: RemoteServiceConnections.Multiplexing })
			await client.handshake({ supportedConnections: RemoteServiceConnections.Multiplexing | RemoteServiceConnections.IpcPipe })
		})
	})

	describe('availabilityChanged', function () {
		it('forwards events across RPC', async function () {
			const client = await getRelayClient()
			const eventRaised = new Promise<BrokeredServicesChangedArgs>(async resolve => {
				client.on('availabilityChanged', args => resolve(args))
			})
			expect(await client.getPipe({ name: 'calc' })).toBeNull()
			innerServer.emit('availabilityChanged', { impactedServices: [{ name: 'calc' }] })
			await eventRaised
		})
	})

	async function getRelayClient() {
		const pair = FullDuplexStream.CreatePair()
		const clientTask = RemoteServiceBroker.connectToMultiplexingDuplex(pair.first)
		const relayTask = MultiplexingRelayServiceBroker.connectToServer(innerServer, pair.second)
		await relayTask
		return await clientTask
	}

	async function getRemoteClientProxy() {
		const pair = FullDuplexStream.CreatePair()
		const relayTask = MultiplexingRelayServiceBroker.connectToServer(innerServer, pair.second)
		const clientMx = await MultiplexingStream.CreateAsync(pair.first)
		const clientChannel = await clientMx.acceptChannelAsync('')
		await relayTask

		const clientBrokerProxy = FrameworkServices.remoteServiceBroker.constructRpc<IRemoteServiceBroker>(clientChannel)
		return clientBrokerProxy
	}

	async function assertCalculatorService(serviceBroker: IServiceBroker | (IServiceBroker & IDisposable)) {
		try {
			const rpc = await serviceBroker.getProxy<ICalculatorService>(calcDescriptorUtf8Http)
			expect(rpc).toBeTruthy()
			expect(await rpc?.add(3, 5)).toEqual(8)
			rpc?.dispose()
		} finally {
			if (IDisposable.is(serviceBroker)) {
				;(serviceBroker as IDisposable).dispose()
			}
		}
	}
})

File: test/Microsoft.ServiceHub.Framework.Tests/MultiplexingRelayServiceBrokerTests.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.IO.Pipelines;
using Microsoft;
using Microsoft.ServiceHub.Framework;
using Microsoft.VisualStudio.Threading;
using Nerdbank.Streams;
using StreamJsonRpc;

public class MultiplexingRelayServiceBrokerTests : TestBase
{
	private IServiceBroker innerServer = new MockServiceBroker();

	public MultiplexingRelayServiceBrokerTests(ITestOutputHelper logger)
		: base(logger)
	{
	}

	[Fact]
	public async Task ConnectToServer_Stream()
	{
		RemoteServiceBroker client = await this.GetRelayClientAsync();
		await this.AssertCalculatorService(client);
	}

	[Fact]
	public async Task ConnectToServer_MultiplexingStream()
	{
		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		Task<RemoteServiceBroker> clientTask = RemoteServiceBroker.ConnectToMultiplexingServerAsync(pair.Item1, this.TimeoutToken);
		Task<MultiplexingStream> serverTask = MultiplexingStream.CreateAsync(pair.Item2, this.TimeoutToken);
		MultiplexingStream serverMxStream = await serverTask;
		MultiplexingStream.Channel relayChannel = await serverMxStream.OfferChannelAsync(string.Empty, this.TimeoutToken);
		var relay = new MultiplexingRelayServiceBroker(this.innerServer, serverMxStream);
		FrameworkServices.RemoteServiceBroker.ConstructRpc(relay, relayChannel);

		await this.AssertCalculatorService(await clientTask);
	}

	[Fact]
	public async Task CancelPendingServiceRequest()
	{
		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		Task<MultiplexingRelayServiceBroker> relayTask = MultiplexingRelayServiceBroker.ConnectToServerAsync(this.innerServer, pair.Item2, this.TimeoutToken);
		MultiplexingStream clientMx = await MultiplexingStream.CreateAsync(pair.Item1, this.TimeoutToken);
		MultiplexingStream.Channel clientChannel = await clientMx.AcceptChannelAsync(string.Empty, this.TimeoutToken);
		await relayTask.WithCancellation(this.TimeoutToken);

		IRemoteServiceBroker clientBrokerProxy = FrameworkServices.RemoteServiceBroker.ConstructRpc<IRemoteServiceBroker>(clientChannel);
		RemoteServiceConnectionInfo connectionInfo = await clientBrokerProxy.RequestServiceChannelAsync(TestServices.Calculator.Moniker, cancellationToken: this.TimeoutToken);
		await clientBrokerProxy.CancelServiceRequestAsync(connectionInfo.RequestId!.Value);

		// This test fails about 50% of the time without this sleep
		Thread.Sleep(50);

		// Assert that the offer for the calculator channel is rescinded.
		Assert.ThrowsAny<InvalidOperationException>(() => clientMx.AcceptChannel(connectionInfo.MultiplexingChannelId!.Value));
	}

	[Fact]
	public async Task RequestNonExistentService()
	{
		RemoteServiceBroker client = await this.GetRelayClientAsync();
		IDuplexPipe? pipe = await client.GetPipeAsync(TestServices.DoesNotExist.Moniker, this.TimeoutToken);
		Assert.Null(pipe);
	}

	[Fact]
	public async Task Handshake_WithoutMultiplexing()
	{
		IRemoteServiceBroker client = await this.GetRemoteClientProxyAsync();
		await Assert.ThrowsAsync<RemoteInvocationException>(() => client.HandshakeAsync(new ServiceBrokerClientMetadata { SupportedConnections = RemoteServiceConnections.IpcPipe }, this.TimeoutToken));
	}

	[Fact]
	public async Task ConnectToServerAsync_Canceled()
	{
		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		await Assert.ThrowsAnyAsync<OperationCanceledException>(() => MultiplexingRelayServiceBroker.ConnectToServerAsync(new MockServiceBroker(), pair.Item1, new CancellationToken(canceled: true)));
		Assert.True(((IDisposableObservable)pair.Item1).IsDisposed);
	}

	[Fact]
	public async Task Ctor_ValidatesInputs()
	{
		Assert.Throws<ArgumentNullException>(() => new MultiplexingRelayServiceBroker(null!, null!));
		Assert.Throws<ArgumentNullException>(() => new MultiplexingRelayServiceBroker(new MockServiceBroker(), null!));

		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		Task<MultiplexingStream> clientTask = MultiplexingStream.CreateAsync(pair.Item1, this.TimeoutToken);
		Task<MultiplexingStream> serverTask = MultiplexingStream.CreateAsync(pair.Item2, this.TimeoutToken);

		MultiplexingStream clientMx = await clientTask;
		Assert.Throws<ArgumentNullException>(() => new MultiplexingRelayServiceBroker(null!, clientMx));
	}

	[Fact]
	public async Task Completion_ClientClosed()
	{
		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		Task<RemoteServiceBroker> clientTask = RemoteServiceBroker.ConnectToMultiplexingServerAsync(pair.Item1, this.TimeoutToken);
		Task<MultiplexingStream> serverTask = MultiplexingStream.CreateAsync(pair.Item2, this.TimeoutToken);
		MultiplexingStream serverMxStream = await serverTask;
		MultiplexingStream.Channel relayChannel = await serverMxStream.OfferChannelAsync(string.Empty, this.TimeoutToken);
		var relay = new MultiplexingRelayServiceBroker(this.innerServer, serverMxStream);
		FrameworkServices.RemoteServiceBroker.ConstructRpc(relay, relayChannel);
		Assert.False(relay.Completion.IsCompleted);

		relayChannel.Dispose();
		await relay.Completion.WithCancellation(this.TimeoutToken);
	}

	private async Task<RemoteServiceBroker> GetRelayClientAsync()
	{
		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		Task<RemoteServiceBroker> clientTask = RemoteServiceBroker.ConnectToMultiplexingServerAsync(pair.Item1, this.TimeoutToken);
		Task<MultiplexingRelayServiceBroker> relayTask = MultiplexingRelayServiceBroker.ConnectToServerAsync(this.innerServer, pair.Item2, this.TimeoutToken);

		await relayTask.WithCancellation(this.TimeoutToken);
		return await clientTask.WithCancellation(this.TimeoutToken);
	}

	private async Task<IRemoteServiceBroker> GetRemoteClientProxyAsync()
	{
		(Stream, Stream) pair = FullDuplexStream.CreatePair();
		Task<MultiplexingRelayServiceBroker> relayTask = MultiplexingRelayServiceBroker.ConnectToServerAsync(this.innerServer, pair.Item2, this.TimeoutToken);
		MultiplexingStream clientMx = await MultiplexingStream.CreateAsync(pair.Item1, this.TimeoutToken);
		MultiplexingStream.Channel clientChannel = await clientMx.AcceptChannelAsync(string.Empty, this.TimeoutToken);
		await relayTask.WithCancellation(this.TimeoutToken);

		IRemoteServiceBroker clientBrokerProxy = FrameworkServices.RemoteServiceBroker.ConstructRpc<IRemoteServiceBroker>(clientChannel);
		return clientBrokerProxy;
	}

	private async Task AssertCalculatorService(IServiceBroker serviceBroker)
	{
		try
		{
			ICalculator? rpc = await serviceBroker.GetProxyAsync<ICalculator>(TestServices.Calculator, this.TimeoutToken);
			Assumes.NotNull(rpc);
			Assert.Equal(8, await rpc.AddAsync(3, 5).WithCancellation(this.TimeoutToken));
			((IDisposable)rpc).Dispose();
		}
		finally
		{
			(serviceBroker as IDisposable)?.Dispose();
		}
	}
}
